app/models/framework/definition/ast/creator.rb

```diff
@@ -4,9 +4,10 @@ module AST
       # This transform exists for the express purpose of creating an AST for generating
       # an ActiveModel class with validations for a Framework
       class Creator < Parslet::Transform
-        rule(string: simple(:s))  { String(s) }
-        rule(decimal: simple(:d)) { BigDecimal(d) }
-        rule(integer: simple(:i)) { Integer(i) }
+        rule(string: simple(:s))           { String(s) }
+        rule(lookup_reference: simple(:r)) { String(r) }
```

Comment:
I was nodding along until now, but between the previous commit ("Values on the RHS of the Lot Number hash should be an identifier, not string") and this one I lost the plot entirely. I might not be the best person to review this, given I haven't been involved in reviewing this before, but I don't understand what makes that thing a `lookup_reference` versus a `pascal_case_identifier`.

Reply:
It's both!

https://github.com/dxw/DataSubmissionServiceAPI/blob/556b691e61c0c944b29663d49e88cc609193fa2b/app/models/framework/definition/parser.rb#L53

We say "this is a `pascal_case_identifier` to say how to parse it, and then we annotate it with `.as(:lookup_reference)` because we know what it is. This rule just transforms a hash, e.g. `{ lookup_reference: 'UnitType' }` into a plain `UnitType` string that the Transpiler consumes via the AST.

Reply:
Honestly I'm often as confused as anyone else 😬 

The annotations `as(:thing)` are pretty much there for convenience when we're transforming the CST into the AST; we added this annotation to help ourselves. I originally had the RHS of the hash as a pascal case identifier (e.g. `MyThing`)  annotated as a string (`value: { string: 'MyThing'} }` but it isn't a string really, it's an identifier pointing to another thing (a lookup list in this case). 

So I was having to do this in the creator to determine if the RHS was a plain value (like `value: 1`) or an annotated value and `to_s` it if it was an annotated value:

`kv.values.first.is_a?(Parslet::Slice) ? kv.values.first.to_s : kv.values.first`

But if we annotate it as a lookup_reference then we can catch it in the creator and just pass the value into the AST, instead of the whole slice.

I realise it's super complex 😬

Reply:
😬 

Why is the thing in the curly brackets `String()`, for both `rule(string: simple(:s))` and `rule(lookup_reference: simple(:r))`?

Reply:
Both of those rules collapse a hash like `{ string: 'A String' }` and `{ lookup_reference: 'SomeLookupReference'}` to just a `String`. In the case of the first, it's turning the type annotation into a real type to be consumed in the AST; in the case of the second it's just removing verbiage like `lookup_reference` and getting rid of the `Parslet::Slice` aspect.  For example, the CST for the field this helps with looks like this:

```
{:field=>"ProductGroup"@633,
    :from=>{:string=>"Vehicle Segment"@652},
    :depends_on=>
     {:dependent_field=>{:string=>"Lot Number"@681},
      :values=>
       {:dictionary=>
         [{:key=>{:string=>"1"@702},
           :value=>{:lookup_reference=>"Lot1Segment"@708}},
          {:key=>{:string=>"2"@727},
           :value=>{:lookup_reference=>"Lot2Segment"@733}},
          {:key=>{:string=>"3"@752},
           :value=>{:lookup_reference=>"Lot3Segment"@758}},
          {:key=>{:string=>"4"@777},
           :value=>{:lookup_reference=>"Lot4Segment"@783}},
          {:key=>{:string=>"5"@802},
           :value=>{:lookup_reference=>"Lot5Segment"@808}},
          {:key=>{:string=>"6"@827},
           :value=>{:lookup_reference=>"Lot6Segment"@833}},
          {:key=>{:string=>"7"@852},
           :value=>{:lookup_reference=>"Lot7Segment"@858}}]}}},
```

You can see the `value` in each is of type `lookup_reference`, and you can see that everything with an `@` is a `Parslet::Slice` (which can *almost* be used interchangeably with a `String` but not quite).

When the rules have finished collapsing that stuff, it looks like this:

```
   {:kind=>:known,
    :field=>"ProductGroup",
    :from=>"Vehicle Segment",
    :depends_on=>
     {:dependent_field=>"Lot Number",
      :values=>
       {"1"=>"Lot1Segment",
        "2"=>"Lot2Segment",
        "3"=>"Lot3Segment",
        "4"=>"Lot4Segment",
        "5"=>"Lot5Segment",
        "6"=>"Lot6Segment",
        "7"=>"Lot7Segment"}}}
```

... which in turn means we can then use `transform_values` on the RHS of the `values` to expand out to full lists. If we didn't use the `lookup_reference` rule it would complicate the AST, not least because it would still contain `Parslet::Slice` objects with their `@10`-style location info, but mostly because there'd be an extra level we don't need (except during the parsing to identify what they are).